File: extensions/shared/src/main/java/app/revanced/extension/youtube/patches/video/requests/PlaylistRequest.java
package app.revanced.extension.youtube.patches.video.requests;

import static app.revanced.extension.shared.patches.spoof.requests.PlayerRoutes.GET_PLAYLIST_PAGE;

import android.annotation.SuppressLint;

import androidx.annotation.GuardedBy;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.SocketTimeoutException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import app.revanced.extension.shared.patches.client.AppClient.ClientType;
import app.revanced.extension.shared.patches.spoof.requests.PlayerRoutes;
import app.revanced.extension.shared.requests.Requester;
import app.revanced.extension.shared.utils.Logger;
import app.revanced.extension.shared.utils.Utils;
import app.revanced.extension.youtube.shared.VideoInformation;

public class PlaylistRequest {

    /**
     * How long to keep fetches until they are expired.
     */
    private static final long CACHE_RETENTION_TIME_MILLISECONDS = 60 * 1000; // 1 Minute

    private static final long MAX_MILLISECONDS_TO_WAIT_FOR_FETCH = 20 * 1000; // 20 seconds

    @GuardedBy("itself")
    private static final Map<String, PlaylistRequest> cache = new HashMap<>();

    @SuppressLint("ObsoleteSdkInt")
    public static void fetchRequestIfNeeded(@Nullable String videoId) {
        Objects.requireNonNull(videoId);
        synchronized (cache) {
            final long now = System.currentTimeMillis();

            cache.values().removeIf(request -> {
                final boolean expired = request.isExpired(now);
                if (expired) Logger.printDebug(() -> "Removing expired stream: " + request.videoId);
                return expired;
            });

            if (!cache.containsKey(videoId)) {
                cache.put(videoId, new PlaylistRequest(videoId));
            }
        }
    }

    @Nullable
    public static PlaylistRequest getRequestForVideoId(@Nullable String videoId) {
        synchronized (cache) {
            return cache.get(videoId);
        }
    }

    private static void handleConnectionError(String toastMessage, @Nullable Exception ex) {
        Logger.printInfo(() -> toastMessage, ex);
    }

    @Nullable
    private static JSONObject send(ClientType clientType, String videoId) {
        Objects.requireNonNull(clientType);
        Objects.requireNonNull(videoId);

        final long startTime = System.currentTimeMillis();
        String clientTypeName = clientType.name();
        Logger.printDebug(() -> "Fetching playlist request for: " + videoId + " using client: " + clientTypeName);

        try {
            HttpURLConnection connection = PlayerRoutes.getPlayerResponseConnectionFromRoute(GET_PLAYLIST_PAGE, clientType);

            String innerTubeBody = String.format(
                    Locale.ENGLISH,
                    PlayerRoutes.createInnertubeBody(clientType, true),
                    videoId,
                    "RD" + videoId
            );
            byte[] requestBody = innerTubeBody.getBytes(StandardCharsets.UTF_8);
            connection.setFixedLengthStreamingMode(requestBody.length);
            connection.getOutputStream().write(requestBody);

            final int responseCode = connection.getResponseCode();
            if (responseCode == 200) return Requester.parseJSONObject(connection);

            handleConnectionError(clientTypeName + " not available with response code: "
                            + responseCode + " message: " + connection.getResponseMessage(),
                    null);
        } catch (SocketTimeoutException ex) {
            handleConnectionError("Connection timeout", ex);
        } catch (IOException ex) {
            handleConnectionError("Network error", ex);
        } catch (Exception ex) {
            Logger.printException(() -> "send failed", ex);
        } finally {
            Logger.printDebug(() -> "video: " + videoId + " took: " + (System.currentTimeMillis() - startTime) + "ms");
        }

        return null;
    }

    private static Boolean fetch(@NonNull String videoId) {
        final ClientType clientType = ClientType.ANDROID_VR;
        final JSONObject playlistJson = send(clientType, videoId);
        if (playlistJson != null) {
            try {
                final JSONObject singleColumnWatchNextResultsJsonObject = playlistJson
                        .getJSONObject("contents")
                        .getJSONObject("singleColumnWatchNextResults");

                if (!singleColumnWatchNextResultsJsonObject.has("playlist")) {
                    return false;
                }

                final JSONObject playlistJsonObject = singleColumnWatchNextResultsJsonObject
                        .getJSONObject("playlist")
                        .getJSONObject("playlist");

                final Object currentStreamObject = playlistJsonObject
                        .getJSONArray("contents")
                        .get(0);

                if (!(currentStreamObject instanceof JSONObject currentStreamJsonObject)) {
                    return false;
                }

                final JSONObject watchEndpointJsonObject = currentStreamJsonObject
                        .getJSONObject("playlistPanelVideoRenderer")
                        .getJSONObject("navigationEndpoint")
                        .getJSONObject("watchEndpoint");

                Logger.printDebug(() -> "watchEndpoint: " + watchEndpointJsonObject);

                return watchEndpointJsonObject.has("playerParams") &&
                        VideoInformation.isMixPlaylistsOpenedByUser(watchEndpointJsonObject.getString("playerParams"));
            } catch (JSONException e) {
                Logger.printDebug(() -> "Fetch failed while processing response data for response: " + playlistJson);
            }
        }

        return false;
    }

    /**
     * Time this instance and the fetch future was created.
     */
    private final long timeFetched;
    private final String videoId;
    private final Future<Boolean> future;

    private PlaylistRequest(String videoId) {
        this.timeFetched = System.currentTimeMillis();
        this.videoId = videoId;
        this.future = Utils.submitOnBackgroundThread(() -> fetch(videoId));
    }

    public boolean isExpired(long now) {
        final long timeSinceCreation = now - timeFetched;
        if (timeSinceCreation > CACHE_RETENTION_TIME_MILLISECONDS) {
            return true;
        }

        // Only expired if the fetch failed (API null response).
        return (fetchCompleted() && getStream() == null);
    }

    /**
     * @return if the fetch call has completed.
     */
    public boolean fetchCompleted() {
        return future.isDone();
    }

    public Boolean getStream() {
        try {
            return future.get(MAX_MILLISECONDS_TO_WAIT_FOR_FETCH, TimeUnit.MILLISECONDS);
        } catch (TimeoutException ex) {
            Logger.printInfo(() -> "getStream timed out", ex);
        } catch (InterruptedException ex) {
            Logger.printException(() -> "getStream interrupted", ex);
            Thread.currentThread().interrupt(); // Restore interrupt status flag.
        } catch (ExecutionException ex) {
            Logger.printException(() -> "getStream failure", ex);
        }

        return null;
    }
}

File: extensions/shared/src/main/java/app/morphe/extension/music/patches/misc/requests/PlaylistRequest.kt
package app.morphe.extension.music.patches.misc.requests

import androidx.annotation.GuardedBy
import app.morphe.extension.shared.innertube.client.YouTubeClient
import app.morphe.extension.shared.innertube.requests.InnerTubeRequestBody.createApplicationRequestBody
import app.morphe.extension.shared.innertube.requests.InnerTubeRequestBody.getInnerTubeResponseConnectionFromRoute
import app.morphe.extension.shared.innertube.requests.InnerTubeRoutes.GET_PLAYLIST_PAGE
import app.morphe.extension.shared.requests.Requester
import app.morphe.extension.shared.settings.AppLanguage
import app.morphe.extension.shared.utils.Logger
import app.morphe.extension.shared.utils.Utils
import app.morphe.extension.shared.utils.Utils.isSDKAbove
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.SocketTimeoutException
import java.util.Objects
import java.util.concurrent.ExecutionException
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

class PlaylistRequest private constructor(
    private val videoId: String,
    private val playlistId: String,
    private val playlistIndex: Int,
) {
    /**
     * Time this instance and the fetch future was created.
     */
    private val timeFetched = System.currentTimeMillis()
    private val future: Future<String> = Utils.submitOnBackgroundThread {
        fetch(
            videoId,
            playlistId,
            playlistIndex,
        )
    }

    fun isExpired(now: Long): Boolean {
        val timeSinceCreation = now - timeFetched
        if (timeSinceCreation > CACHE_RETENTION_TIME_MILLISECONDS) {
            return true
        }

        // Only expired if the fetch failed (API null response).
        return (fetchCompleted() && songId.isEmpty())
    }

    /**
     * @return if the fetch call has completed.
     */
    fun fetchCompleted(): Boolean {
        return future.isDone
    }

    val songId: String
        get() {
            try {
                return future[MAX_MILLISECONDS_TO_WAIT_FOR_FETCH, TimeUnit.MILLISECONDS]
            } catch (ex: TimeoutException) {
                Logger.printInfo(
                    { "getSongId timed out" },
                    ex
                )
            } catch (ex: InterruptedException) {
                Logger.printException(
                    { "getSongId interrupted" },
                    ex
                )
                Thread.currentThread().interrupt() // Restore interrupt status flag.
            } catch (ex: ExecutionException) {
                Logger.printException(
                    { "getSongId failure" },
                    ex
                )
            }

            return ""
        }

    companion object {
        /**
         * How long to keep fetches until they are expired.
         */
        private const val CACHE_RETENTION_TIME_MILLISECONDS = 60 * 1000L // 1 Minute

        private const val MAX_MILLISECONDS_TO_WAIT_FOR_FETCH = 10 * 1000L // 10 seconds

        @GuardedBy("itself")
        private val cache: MutableMap<String, PlaylistRequest> = HashMap()

        @JvmStatic
        fun fetchRequestIfNeeded(
            videoId: String,
            playlistId: String,
            playlistIndex: Int,
        ) {
            Objects.requireNonNull(videoId)
            synchronized(cache) {
                val now = System.currentTimeMillis()
                if (isSDKAbove(24)) {
                    cache.values.removeIf { request ->
                        val expired = request.isExpired(now)
                        if (expired) Logger.printDebug { "Removing expired song id: " + request.videoId }
                        expired
                    }
                } else {
                    val itr = cache.entries.iterator()
                    while (itr.hasNext()) {
                        val request = itr.next().value
                        if (request.isExpired(now)) {
                            Logger.printDebug { "Removing expired song id: " + request.videoId }
                            itr.remove()
                        }
                    }
                }

                if (!cache.containsKey(videoId)) {
                    cache[videoId] = PlaylistRequest(
                        videoId,
                        playlistId,
                        playlistIndex,
                    )
                }
            }
        }

        @JvmStatic
        fun getRequestForVideoId(videoId: String): PlaylistRequest? {
            synchronized(cache) {
                return cache[videoId]
            }
        }

        private fun handleConnectionError(toastMessage: String, ex: Exception?) {
            Logger.printInfo({ toastMessage }, ex)
        }

        private fun sendRequest(
            videoId: String,
            playlistId: String,
        ): JSONObject? {
            Objects.requireNonNull(videoId)

            val startTime = System.currentTimeMillis()
            val clientType = YouTubeClient.ClientType.ANDROID_VR_NO_AUTH
            val clientTypeName = clientType.name
            Logger.printDebug { "Fetching playlist request for: $videoId, using client: $clientTypeName" }

            try {
                val connection = getInnerTubeResponseConnectionFromRoute(
                    GET_PLAYLIST_PAGE,
                    clientType
                )

                /**
                 * For some reason, the tracks in Top Songs have the playlistId of the album:
                 * [ReVanced_Extended#2835](https://github.com/inotia00/ReVanced_Extended/issues/2835)
                 *
                 * We can work around this issue by checking the playlist title in the response.
                 * Tracks played from an album have a playlist title that starts with 'Album',
                 * Tracks played from Top Songs have a playlist title that starts with 'Song'.
                 *
                 * By default, the playlist title follows the app language,
                 * So we can work around this by setting the language to English when sending the request.
                 */
                val requestBody =
                    createApplicationRequestBody(
                        clientType = clientType,
                        videoId = videoId,
                        playlistId = playlistId,
                        language = AppLanguage.EN.language,
                    )

                connection.setFixedLengthStreamingMode(requestBody.size)
                connection.outputStream.write(requestBody)

                val responseCode = connection.responseCode
                if (responseCode == 200) return Requester.parseJSONObject(connection)

                handleConnectionError(
                    (clientTypeName + " not available with response code: "
                            + responseCode + " message: " + connection.responseMessage),
                    null
                )
            } catch (ex: SocketTimeoutException) {
                handleConnectionError("Connection timeout", ex)
            } catch (ex: IOException) {
                handleConnectionError("Network error", ex)
            } catch (ex: Exception) {
                Logger.printException({ "sendRequest failed" }, ex)
            } finally {
                Logger.printDebug { "video: " + videoId + " took: " + (System.currentTimeMillis() - startTime) + "ms" }
            }

            return null
        }

        private fun parseResponse(playlistJson: JSONObject, playlistIndex: Int): String {
            try {
                val singleColumnWatchNextResultsJsonObject: JSONObject =
                    playlistJson
                        .getJSONObject("contents")
                        .getJSONObject("singleColumnWatchNextResults")

                if (singleColumnWatchNextResultsJsonObject.has("playlist")) {
                    val playlistJsonObject: JSONObject? =
                        singleColumnWatchNextResultsJsonObject
                            .getJSONObject("playlist")
                            .getJSONObject("playlist")

                    val playlistTitle = playlistJsonObject
                        ?.getString("title") + ""

                    if (playlistTitle.startsWith("Album")) {
                        val currentStreamJsonObject = playlistJsonObject
                            ?.getJSONArray("contents")
                            ?.get(playlistIndex)

                        if (currentStreamJsonObject is JSONObject) {
                            val watchEndpointJsonObject: JSONObject? =
                                currentStreamJsonObject
                                    .getJSONObject("playlistPanelVideoRenderer")
                                    .getJSONObject("navigationEndpoint")
                                    .getJSONObject("watchEndpoint")

                            return watchEndpointJsonObject?.getString("videoId") + ""
                        }
                    }
                }
            } catch (e: JSONException) {
                Logger.printException(
                    { "Fetch failed while processing response data for response: $playlistJson" },
                    e
                )
            }

            return ""
        }

        private fun fetch(
            videoId: String,
            playlistId: String,
            playlistIndex: Int,
        ): String {
            val playlistJson = sendRequest(
                videoId,
                playlistId,
            )
            if (playlistJson != null) {
                return parseResponse(playlistJson, playlistIndex)
            }

            return ""
        }
    }
}
